File: src/types/specialAbility/ProtectiveWardingCircleSpecialAbility.ts
import {
  Entity,
  IncludeIdentifier,
  Integer,
  Object,
  Optional,
  Required,
  String,
} from "tsondb/schema/def"
import {
  ap_value,
  levels,
  maximum,
  name,
  name_in_library,
  select_options,
} from "../_Activatable.js"
import { GeneralPrerequisites } from "../_Prerequisite.js"
import { NestedLocaleMap } from "../Locale.js"
import { Errata } from "../source/_Erratum.js"
import { src } from "../source/_PublicationRef.js"

export const ProtectiveWardingCircleSpecialAbility = Entity(import.meta.url, {
  name: "ProtectiveWardingCircleSpecialAbility",
  namePlural: "ProtectiveWardingCircleSpecialAbilities",
  type: () =>
    Object({
      levels,
      select_options,
      maximum,
      prerequisites: Optional({
        type: IncludeIdentifier(GeneralPrerequisites),
      }),
      cost: Required({
        comment: "The cost in AE.",
        type: Integer({ minimum: 0 }),
      }),
      ap_value,
      src,
      translations: NestedLocaleMap(
        Required,
        "ProtectiveWardingCircleSpecialAbilityTranslation",
        Object({
          name,
          name_in_library,
          protective_circle: Required({
            comment: "The rules for the protective circle variant.",
            type: String({ minLength: 1, isMarkdown: true }),
          }),
          warding_circle: Required({
            comment: "The rules for the warding circle variant.",
            type: String({ minLength: 1, isMarkdown: true }),
          }),
          errata: Optional({
            type: IncludeIdentifier(Errata),
          }),
        })
      ),
    }),
  displayName: {},
})

File: Sources/OptolithDatabaseSchema/Types/specialAbility/ProtectiveWardingCircleSpecialAbility.swift
import FileDB

@Model
public struct ProtectiveWardingCircleSpecialAbility {
      levels,
      select_options,
      maximum,
      prerequisites: Optional({
        type: IncludeIdentifier(GeneralPrerequisites),
      }),

  /// The cost in AE.
  let ae_cost: Integer({ minimum: 0 })
      ap_value,

    /// The publications where you can find the entry.
    let src: PublicationRefs

    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship
    let translations: [String: Translation]

    struct Translation { // ProtectiveWardingCircleSpecialAbilityTranslation
          name,
          name_in_library,

        /// The rules for the protective circle variant.
        let protective_circle: String({ minLength: 1, isMarkdown: true })

        /// The rules for the warding circle variant.
        let warding_circle: String({ minLength: 1, isMarkdown: true })

        let errata: Errata?
    }
}
